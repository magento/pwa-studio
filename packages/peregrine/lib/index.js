import * as RestApi from './RestApi';
import * as Util from './util';

// hooks
export { useEventListener } from './hooks/useEventListener';
export { useCarousel } from './hooks/useCarousel';
export { useDropdown } from './hooks/useDropdown';
export { usePagination } from './hooks/usePagination';
export { useRestApi } from './hooks/useRestApi';
export { useRestResponse } from './hooks/useRestResponse';
export { useScrollLock } from './hooks/useScrollLock';
export { useSearchParam } from './hooks/useSearchParam';
export { useSort } from './hooks/useSort';
export { useTypePolicies } from './hooks/useTypePolicies';

export {
    WindowSizeContextProvider,
    useWindowSize
} from './hooks/useWindowSize';
export { getToastId, useToasts, ToastContextProvider } from './Toasts';

// store
export { enhancer, reducers } from './store';

// components
export { default as ContainerChild } from './ContainerChild';
export { default as List, Items, Item } from './List';
export { default as Price } from './Price';
export { default as Router } from './Router';
export {
    default as PeregrineContextProvider
} from './PeregrineContextProvider';

// misc
export { RestApi };
export { Util };
export { default as createTestInstance } from './util/createTestInstance';
