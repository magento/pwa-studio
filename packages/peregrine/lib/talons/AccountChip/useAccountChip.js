import { useMemo } from 'react';
import { useQuery } from '@apollo/client';
import { useUserContext } from '@magento/peregrine/lib/context/user';

/**
 * The useAccountChip talon supports the AccountChip component.
 *
 * @param {GraphQLAST} props.queries.getCustomerDetailsQuery
 *
 * @returns {Object} talonProps
 * @returns {Object} talonProps.currentUser - Details about the currently signed-in user.
 * @returns {Bool}   talonProps.isLoadingUserName - Indicates when we know there is a
 *  user signed in, but we don't yet have their name.
 * @returns {Bool}   talonProps.isUserSignedIn - Indicates whether we have a signed-in user.
 */
export const useAccountChip = props => {
    const {
        queries: { getCustomerDetailsQuery }
    } = props;
    const [{ isSignedIn: isUserSignedIn }] = useUserContext();

    const { data } = useQuery(getCustomerDetailsQuery, {
        fetchPolicy: 'cache-and-network',
        nextFetchPolicy: 'cache-first',
        skip: !isUserSignedIn
    });

    const currentUser = useMemo(() => {
        return (data && data.customer) || null;
    }, [data]);

    return {
        currentUser,
        isLoadingUserName: isUserSignedIn && !currentUser,
        isUserSignedIn
    };
};
