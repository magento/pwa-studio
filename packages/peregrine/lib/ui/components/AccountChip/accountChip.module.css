.root {
    composes: gap-x-2xs from global;
    composes: grid from global;
    composes: grid-cols-autoFirst from global;
    composes: grid-flow-col from global;
    composes: items-center from global;
}

.loader {
    composes: indicator from '../LoadingIndicator/indicator.module.css';
}
