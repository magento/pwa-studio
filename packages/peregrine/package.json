{
  "name": "@magento/peregrine",
  "version": "12.1.0",
  "publishConfig": {
    "access": "public"
  },
  "description": "The core runtime of Magento PWA",
  "main": "lib/index.js",
  "scripts": {
    "clean": " "
  },
  "repository": "github:magento/pwa-studio",
  "author": "Magento Commerce",
  "license": "(OSL-3.0 OR AFL-3.0)",
  "bugs": {
    "url": "https://github.com/magento/pwa-studio/issues"
  },
  "homepage": "https://github.com/magento/pwa-studio/tree/master/packages/peregrine#readme",
  "dependencies": {
    "fast-glob": "~3.2.4"
  },
  "devDependencies": {
    "@apollo/client": "~3.4.0",
    "@magento/eslint-config": "~1.5.0",
    "@testing-library/react-hooks": "~5.0.3",
    "intl": "~1.2.5",
    "intl-locales-supported": "~1.8.12",
    "react": "~17.0.1",
    "react-dom": "~17.0.1",
    "react-intl": "~5.20.0",
    "react-redux": "~7.2.2",
    "react-router-dom": "~5.2.0",
    "react-test-renderer": "~17.0.1",
    "redux": "~4.0.4",
    "redux-actions": "~2.6.5",
    "wait-for-expect": "~1.2.0"
  },
  "peerDependencies": {
    "@apollo/client": "~3.4.0",
    "@babel/runtime": "~7.15.3",
    "informed": "~3.29.0",
    "react": "~17.0.1",
    "react-dom": "~17.0.1",
    "react-intl": "~5.20.0",
    "react-redux": "~7.2.2",
    "react-router-dom": "~5.2.0",
    "redux": "~4.0.4",
    "redux-actions": "~2.6.5"
  },
  "engines": {
    "node": ">=14.x",
    "yarn": ">=1.12.0"
  },
  "pwa-studio": {
    "targets": {
      "declare": "./lib/targets/peregrine-declare",
      "intercept": "./lib/targets/peregrine-intercept"
    }
  },
  "module": "lib/index.js",
  "jsnext:main": "lib/index.js",
  "es2015": "lib/index.js",
  "sideEffects": false
}
