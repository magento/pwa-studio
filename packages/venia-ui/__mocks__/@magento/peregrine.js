import React from 'react';
import { BrowserPersistence } from './util/simplePersistence';

const peregrine = jest.requireActual('../../../peregrine/lib');

// re-exports
const RestApi = {
    Magento2: {
        request: jest.fn()
    }
};

const Util = { BrowserPersistence };

// hooks
const useEventListener = jest.fn(peregrine.useEventListener);
const useDropdown = jest.fn(peregrine.useDropdown);
const useSearchParam = jest.fn(peregrine.useSearchParam);
const useWindowSize = jest.fn(peregrine.useWindowSize);
const useToasts = jest.fn(peregrine.useToasts);
const useTypePolicies = jest.fn(peregrine.useTypePolicies);
// components

/**
 * the Price component from @magento/peregrine
 * has browser-specific functionality and cannot
 * currently by rendered in the test environment
 */
const Price = jest.fn().mockReturnValue(<div />);

module.exports = {
    ...peregrine,
    Price,
    RestApi,
    Util,
    useEventListener,
    useDropdown,
    useSearchParam,
    useWindowSize,
    useToasts,
    useTypePolicies
};
