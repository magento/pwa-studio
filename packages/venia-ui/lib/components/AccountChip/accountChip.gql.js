import { gql } from '@apollo/client';

export const GET_CUSTOMER_DETAILS = gql`
    query accountChipQuery {
        # eslint-disable-next-line @graphql-eslint/require-id-when-available
        customer {
            firstname
        }
    }
`;
