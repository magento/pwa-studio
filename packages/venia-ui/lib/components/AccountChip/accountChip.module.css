.root {
    align-items: center;
    column-gap: 0.5rem;
    display: grid;
    grid-auto-flow: column;
    grid-template-columns: auto 1fr;
}

.loader {
    composes: indicator from '../LoadingIndicator/indicator.module.css';
}
