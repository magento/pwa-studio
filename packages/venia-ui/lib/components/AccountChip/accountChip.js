import React from 'react';
import { useIntl } from 'react-intl';
import { bool, shape, string } from 'prop-types';
import { Loader, User as AccountIcon } from 'react-feather';

import { useAccountChip } from '@magento/peregrine/lib/talons/AccountChip/useAccountChip';
import { useStyle } from '@magento/venia-ui/lib/classify';

import Icon from '../Icon';
import defaultClasses from './accountChip.module.css';
import { GET_CUSTOMER_DETAILS } from './accountChip.gql';

/**
 * The AccountChip component shows an icon next to some text.
 * Sometimes the text is static, sometimes it is dynamic based on the user's name,
 * and it can also be a loading icon to indicate that we're fetching the user's name.
 *
 * @param {Object} props
 * @param {Object} props.classes - CSS classes to override element styles.
 * @param {String} props.fallbackText - The text to display when the user is not signed in
 *  or when we're loading details but don't want to show a loading icon.
 * @param {Boolean} props.shouldIndicateLoading - Whether we should show a loading icon or
 *  not when the user is signed in but we don't have their details (name) yet.
 */
const AccountChip = props => {
    const { fallbackText, shouldIndicateLoading } = props;

    const talonProps = useAccountChip({
        queries: {
            getCustomerDetailsQuery: GET_CUSTOMER_DETAILS
        }
    });
    const { currentUser, isLoadingUserName, isUserSignedIn } = talonProps;

    const classes = useStyle(defaultClasses, props.classes);
    const { formatMessage } = useIntl();

    let chipText;
    if (!isUserSignedIn) {
        chipText = fallbackText;
    } else {
        if (!isLoadingUserName) {
            chipText = formatMessage(
                { id: 'accountChip.chipText', defaultMessage: 'Hi, {name}' },
                { name: currentUser.firstname }
            );
        } else if (shouldIndicateLoading) {
            chipText = <Icon classes={{ icon: classes.loader }} src={Loader} />;
        } else {
            chipText = fallbackText;
        }
    }

    return (
        <span className={classes.root}>
            <Icon src={AccountIcon} />
            <span className={classes.text}>{chipText}</span>
        </span>
    );
};

export default AccountChip;

AccountChip.propTypes = {
    classes: shape({
        root: string,
        loader: string,
        text: string
    }),
    fallbackText: string,
    shouldIndicateLoading: bool
};

AccountChip.defaultProps = {
    fallbackText: 'Account',
    shouldIndicateLoading: false
};
