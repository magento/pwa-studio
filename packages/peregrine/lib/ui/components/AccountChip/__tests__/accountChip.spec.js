import React from 'react';
import { createTestInstance } from '@magento/peregrine';

import { useAccountChip } from '@magento/peregrine/lib/talons/AccountChip/useAccountChip';

import AccountChip from '../accountChip';

jest.mock('../../../classify');

jest.mock('@magento/peregrine/lib/talons/AccountChip/useAccountChip', () => {
    const useAccountChipTalon = jest.requireActual(
        '@magento/peregrine/lib/talons/AccountChip/useAccountChip'
    );
    const spy = jest.spyOn(useAccountChipTalon, 'useAccountChip');

    return Object.assign(useAccountChipTalon, { useAccountChip: spy });
});

const talonProps = {
    currentUser: {},
    isLoadingUserName: false,
    isUserSignedIn: false
};

test('it renders the default fallback correctly', () => {
    // Arrange.
    useAccountChip.mockReturnValueOnce(talonProps);

    // Act.
    const { root } = createTestInstance(<AccountChip />);

    // Assert.
    expect(
        root.find(({ children }) =>
            children.includes(AccountChip.defaultProps.fallbackText)
        )
    ).toBeTruthy();
});

test('it renders a prop fallback correctly', () => {
    // Arrange.
    useAccountChip.mockReturnValueOnce(talonProps);
    const props = {
        fallbackText: 'unit test'
    };

    // Act.
    const { root } = createTestInstance(<AccountChip {...props} />);

    // Assert.
    expect(
        root.find(({ children }) => children.includes(props.fallbackText))
    ).toBeTruthy();
});

test('it renders a user greeting correctly', () => {
    // Arrange.
    const myTalonProps = {
        ...talonProps,
        currentUser: { firstname: 'Unit Tester' },
        isLoadingUserName: false,
        isUserSignedIn: true
    };
    useAccountChip.mockReturnValueOnce(myTalonProps);

    // Act.
    const { root } = createTestInstance(<AccountChip />);

    // Assert.
    expect(
        root.find(({ children }) => {
            return children.includes(
                `Hi, ${myTalonProps.currentUser.firstname}`
            );
        })
    ).toBeTruthy();
});

test('it renders a loading indicator when appropriate', () => {
    // Arrange.
    const myTalonProps = {
        ...talonProps,
        currentUser: {},
        isLoadingUserName: true,
        isUserSignedIn: true
    };
    useAccountChip.mockReturnValueOnce(myTalonProps);

    const props = {
        shouldIndicateLoading: true
    };

    // Act.
    const instance = createTestInstance(<AccountChip {...props} />);

    // Assert.
    expect(instance.toJSON()).toMatchSnapshot();
});

test('it renders fallback text when loading but instructed not to show loading indicator', () => {
    // Arrange.
    const myTalonProps = {
        ...talonProps,
        currentUser: {},
        isLoadingUserName: true,
        isUserSignedIn: true
    };
    useAccountChip.mockReturnValueOnce(myTalonProps);

    const props = {
        fallbackText: 'Unit Test',
        shouldIndicateLoading: false
    };

    // Act.
    const { root } = createTestInstance(<AccountChip {...props} />);

    // Assert.
    expect(
        root.find(({ children }) => {
            return children.includes(props.fallbackText);
        })
    ).toBeTruthy();
});
