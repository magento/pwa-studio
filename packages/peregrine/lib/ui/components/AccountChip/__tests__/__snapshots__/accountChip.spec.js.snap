// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders a loading indicator when appropriate 1`] = `
<span
  className="root"
>
  <span
    className="root"
  >
    <svg
      className="icon"
      fill="none"
      height={24}
      stroke="currentColor"
      strokeLinecap="round"
      strokeLinejoin="round"
      strokeWidth="2"
      viewBox="0 0 24 24"
      width={24}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M20 21v-2a4 4 0 0 0-4-4H8a4 4 0 0 0-4 4v2"
      />
      <circle
        cx="12"
        cy="7"
        r="4"
      />
    </svg>
  </span>
  <span
    className="text"
  >
    <span
      className="root"
    >
      <svg
        className="icon"
        fill="none"
        height={24}
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        viewBox="0 0 24 24"
        width={24}
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="12"
          x2="12"
          y1="2"
          y2="6"
        />
        <line
          x1="12"
          x2="12"
          y1="18"
          y2="22"
        />
        <line
          x1="4.93"
          x2="7.76"
          y1="4.93"
          y2="7.76"
        />
        <line
          x1="16.24"
          x2="19.07"
          y1="16.24"
          y2="19.07"
        />
        <line
          x1="2"
          x2="6"
          y1="12"
          y2="12"
        />
        <line
          x1="18"
          x2="22"
          y1="12"
          y2="12"
        />
        <line
          x1="4.93"
          x2="7.76"
          y1="19.07"
          y2="16.24"
        />
        <line
          x1="16.24"
          x2="19.07"
          y1="7.76"
          y2="4.93"
        />
      </svg>
    </span>
  </span>
</span>
`;
